Binder (#1)
ome/omero-guide-trackmate: docs/trackmate_scripting.rst

-  How to create a TrackMate model using its API.

-  How to save the tracks as polygon ROIs in OMERO.

Setup
-----

- Install Fiji on the local machine with the OMERO.insight-ij plugin. The installation instructions can be found at `here <https://omero-guides.readthedocs.io/en/latest/fiji/docs/installation.html>`_.

Resources
---------

-  We will use a timelapse image available at `DV/iain/438CTR <https://downloads.openmicroscopy.org/images/DV/iain/438CTR/>`_

-  Script: Groovy script for tracking timelapse images - :download:`tracking.groovy <../scripts/tracking.groovy>`.

Step-by-Step
------------

In this section, we go through the steps required to analyze the data.
The script used in this document is :download:`tracking.groovy <../scripts/tracking.groovy>`. The script follows similar steps than the ones used via the User Interface, see :doc:`trackmate_ui`.

One advantage of the scripting approach is that we can save the generated tracks as 
ROIs in an OMERO server.

#. Launch Fiji.

#. Open ``File > New > Script...``.

#. Select ``Groovy`` as a language.

#. Copy the content of :download:`tracking.groovy <../scripts/tracking.groovy>` in the text area.

> #. A dialog will pop up. Enter the credentials to connect to the server and select an Image.

Reviewer: ```suggestion
#. A dialog will pop up. Enter the credentials to connect to the server. Note that the server address should be in form ``wss://your-server-address/omero-ws``  or similar. This means the usage of websockets is necessary. The ending of the address (``omero-ws`` in this example) might be different for your server. 

#. Select an Image from OMERO and enter the ID of the image into the dialog in Fiji.
```

Author: The usage of websocket is not necessary.

Reviewer: But when I try it without websocket, I get  (repeatedly, 3 times)

From main console popup in Fiji

```
Parsing credentials
Logging in
-! 18/06/20 16:24:16:029 warning: SciJava-49af9ddb-Thread-0: null - createSession retry: 1
-! 18/06/20 16:24:21:658 warning: SciJava-49af9ddb-Thread-0: null - createSession retry: 2
```

From the scipt editor console (after a while) - expand details to see error

<details>

```
omero.ClientError: Obtained null object proxy
	at omero.client.createSession(client.java:837)
	at loci.ome.io.OmeroReader.initFile(OmeroReader.java:292)
	at loci.formats.FormatReader.setId(FormatReader.java:1389)
	at loci.plugins.in.ImportProcess.initializeFile(ImportProcess.java:499)
	at loci.plugins.in.ImportProcess.execute(ImportProcess.java:142)
	at loci.plugins.in.Importer.showDialogs(Importer.java:140)
	at loci.plugins.in.Importer.run(Importer.java:76)
	at loci.plugins.LociImporter.run(LociImporter.java:78)
	at ij.IJ.runUserPlugIn(IJ.java:230)
	at ij.IJ.runPlugIn(IJ.java:193)
	at ij.IJ.runPlugIn(IJ.java:182)
	at ij.IJ$runPlugIn.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
	at Script1.open_image_plus(Script1.groovy:124)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1218)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1027)
	at groovy.lang.DelegatingMetaClass.invokeMethod(DelegatingMetaClass.java:151)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine$2.invokeMethod(GroovyScriptEngine.java:283)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:69)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:52)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:154)
	at Script1.run(Script1.groovy:241)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:303)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:122)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.scijava.script.ScriptModule.run(ScriptModule.java:160)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:168)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:66)
	at org.scijava.thread.DefaultThreadService$3.call(DefaultThreadService.java:238)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

</details>

Whereas with websocket it repeatedly works.

Author: you are right if it is b/c I have to specify the port for BF plugin due to https://github.com/ome/omero-blitz/issues/92

The core difference of this approach is that the generated track **can** be saved as
OMERO ROIs.

Description
-----------

First, we will show how to use the TrackMate API and the Scripting editor of Fiji.

We will show:

-  How to connect to OMERO using the JAVA API.

-  How to retrieve an image.

-  How to open the image using Bio-Formats.

-  How to create a TrackMate model using its API.

-  How to save the tracks as polygon ROIs in OMERO.

Setup
-----

- Install Fiji on the local machine with the OMERO.insight-ij plugin. The installation instructions can be found at `here <https://omero-guides.readthedocs.io/en/latest/fiji/docs/installation.html>`_.

Resources
---------

-  We will use a timelapse image available at `DV/iain/438CTR <https://downloads.openmicroscopy.org/images/DV/iain/438CTR/>`_

-  Script: Groovy script for tracking timelapse images - :download:`tracking.groovy <../scripts/tracking.groovy>`.

Step-by-Step
------------

In this section, we go through the steps required to analyze the data.
The script used in this document is :download:`tracking.groovy <../scripts/tracking.groovy>`.

> A 'Tracking' Jupyter notebook is also available in the notebooks section.

Reviewer: Now it is not clear where such section is and how to access it. I guess the section here is missing aka Option 2 in the ``python-Fiji`` PR.

Author: I did not want to include how to run the notebook from that page.
So I will remove that part

One advantage of the scripting approach is that we can save the generated tracks as 
ROIs in an OMERO server.

#. Launch Fiji.

#. Open ``File > New > Script...``.

#. Select ``Groovy`` as a language.

#. Copy the content of :download:`tracking.groovy <../scripts/tracking.groovy>` in the text area.

> #. Enter the credentials to connect to the server and select an Image.

Reviewer: ```suggestion
#. A dialog will pop up. Enter the credentials to connect to the server and select an Image.
```

Reviewer: But here is a mistake in the order of appearance I guess. That dialog can only appear after you clicked Run (next step below.)

#. Click Run.

> #. Go back to OMERO.web to visualize the tracks.

Reviewer: ```suggestion
#. Go back to OMERO.web to visualize the tracks. Double-click on the image in OMERO.web to open it in OMERO.iviewer.
```

Analyze OMERO timelapse images using the TrackMate API
======================================================

In this example we open an image stored in an OMERO server and use the `TrackMate <https://imagej.net/TrackMate>`_ API to analyze it.

> The core difference of this approach is that the generated track **can** be saved as

Reviewer: ```suggestion
One of the advantages of this approach over the User interface workflow is that the generated track **can** be saved as
```

OMERO ROIs.

Description
-----------

First, we will show how to use the TrackMate API and the Scripting editor of Fiji.

We will show:

-  How to connect to OMERO using the JAVA API.

-  How to retrieve an image.

-  How to open the image using Bio-Formats.

-  How to create a TrackMate model using its API.

> -  How to save the tracks as polygons ROIs in OMERO.

Reviewer: ```suggestion
-  How to save the tracks as polygon ROIs in OMERO.
```
